### PR TITLE
feat: gcp sql database support

### PR DESCRIPTION
**Pull request by HomelessDinosaur**

### CODE REVIEW

File: cloud/gcp/deploy/api.go [hunk 1]
```diff
@@ -153,24 +144,22 @@ func (p *NitricGcpPulumiProvider) Api(ctx *pulumi.Context, parent pulumi.Resourc
...
-			p.Put = gcpOperation(name, p.Put, naps, timeouts)
-			p.Delete = gcpOperation(name, p.Delete, naps, timeouts)
-			p.Options = gcpOperation(name, p.Options, naps, timeouts)
+			p.Get = gcpOperation(name, p.Get, naps)
+			p.Post = gcpOperation(name, p.Post, naps)
+			p.Patch = gcpOperation(name, p.Patch, naps)
+			p.Put = gcpOperation(name, p.Put, naps)
+			p.Delete = gcpOperation(name, p.Delete, naps)
+			p.Options = gcpOperation(name, p.Options, naps)
 			v2doc.Paths[k] = p
```
Comment (tjholm) [inline cloud/gcp/deploy/api.go:163]: This clobbers a fix for gateway route timeouts not matching service timeouts

File: cloud/gcp/deploy/api.go [hunk 2]
```diff
@@ -277,7 +266,7 @@ func keepOperation(opExt map[string]interface{}) (string, bool) {
 	return name, true
 }
 
-func gcpOperation(apiName string, op *openapi2.Operation, urls map[string]string, timeouts map[string]int) *openapi2.Operation {
+func gcpOperation(apiName string, op *openapi2.Operation, urls map[string]string) *openapi2.Operation {
```
Comment (tjholm) [inline cloud/gcp/deploy/api.go:269]: fix reversion of gateway route timeout fix.

File: cloud/gcp/deploy/api.go [hunk 3]
```diff
@@ -309,11 +298,10 @@ func gcpOperation(apiName string, op *openapi2.Operation, urls map[string]string
 		}
 	}
 
-	op.Extensions["x-google-backend"] = map[string]any{
+	op.Extensions["x-google-backend"] = map[string]string{
 		// Append the name of the target origin api gateway to the target address
 		"address":          fmt.Sprintf("%s/x-nitric-api/%s", urls[name], apiName),
 		"path_translation": "APPEND_PATH_TO_ADDRESS",
-		"deadline":         timeouts[name],
```
Comment (tjholm) [inline cloud/gcp/deploy/api.go:316]: Fix reversion of gateway -> service timeout fixes.

**Review by tjholm — CHANGES_REQUESTED** [review]
Changes that fix a gateway to service timeout mismatch have been reverted in the PR.

**Review by tjholm — APPROVED** [review]

**Review by davemooreuws — CHANGES_REQUESTED** [review]
First deploy works well.

But next ones don't apply the migrations at the correct time, the cloud run builds are triggered before the new image is available. I used prisma to test.

Also might need to include similar dependencies for deploying services after migrations, like these: https://github.com/nitrictech/nitric/pull/627

**Review by jyecusch — APPROVED** [review]

File: cloud/gcp/deploy/sql.go [hunk 4]
```diff
@@ -17,13 +17,189 @@
...
+
+	ID pulumi.StringOutput
+}
+
 func (a *NitricGcpPulumiProvider) SqlDatabase(ctx *pulumi.Context, parent pulumi.Resource, name string, config *deploymentspb.SqlDatabase) error {
-	return fmt.Errorf("Sql databases are unimplemented in the nitric GCP provider")
+	_, err := sql.NewDatabase(ctx, name, &sql.DatabaseArgs{
+		Name:           pulumi.String(name),
+		Instance:       a.masterDb.Name,
+		DeletionPolicy: pulumi.String(a.GcpConfig.Databases[name].DeletionPolicy),
```
Comment (davemooreuws) [inline cloud/gcp/deploy/sql.go:48]: I seem to be getting a nil pointer error here.

```
iagnostics:
  pulumi:pulumi:Stack (nitric-prisma-final-nitric-prisma-final-gcpsql):
    error: an unhandled error occurred: 1 error occurred:
        * recovered panic: runtime error: invalid memory address or nil pointer dereference
     Stack: goroutine 90 [running]:
    runtime/debug.Stack()
        C:/Program Files/Go/src/runtime/debug/stack.go:24 +0x5e
    github.com/nitrictech/nitric/cloud/common/deploy/provider.(*PulumiProviderServer).Up.createPulumiProgramForNitricProvider.func2.1()
        C:/projects/nitric/nitric/cloud/common/deploy/provider/pulumi.go:80 +0x3d
    panic({0x2fb66a0?, 0x86e6000?})
        C:/Program Files/Go/src/runtime/panic.go:770 +0x132
    github.com/nitrictech/nitric/cloud/gcp/deploy.(*NitricGcpPulumiProvider).SqlDatabase(0xc00034a9c0, 0xc0007c5440, {0x43f1380, 0xc0007fcdd0},
{0xc0001aa070, 0x5}, 0xc0005deac0)
        C:/projects/nitric/nitric/cloud/gcp/deploy/sql.go:48 +0x1ff
```

**Review by davemooreuws — CHANGES_REQUESTED** [review]
Issue with deletion policy found, see comments.

File: cloud/gcp/deploy/sql.go (hunk 4 above)
Comment (davemooreuws) [reply cloud/gcp/deploy/sql.go:48]: Suggestion, check for a.GcpConfig.Databases[name] != nil then apply the deletion policy.

File: cloud/gcp/runtime/sql/cloud_sql.go
Comment (davemooreuws) [inline cloud/gcp/runtime/sql/cloud_sql.go:1]: This needs to be added to the server.go

**Comment by davemooreuws** [comment]
Last issue is making sure services are created after the cloud build run ends.

**Comment by tjholm** [comment]
Fixes for existing issues implemented up/down is smooth now. Should be good to release.

**Pull request merged**